Terabox backend

x1arch opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?
This is Terabox storage implementation. 

This integrations is not official, because official integration required special `client id` and `client secret` which only can be provided by Terabox. I made a request for this, but did not got a response, yet. If Terabox will provide the keys for official integration, will update the integration.

This is alpha version, not finished, not all scenarios are tested.

Issues: #7496; #8507; #8356; #7244; #6904; #6122 

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

BenjiThatFoxGuy commented:
I tried the fork and it fails to list files in a mount and/or in rclone ls if i specify / for listing the whole terabox account, and i have to do rclone ls teraboxbackend:.
very weird (yes i know it's a WIP, just reporting what i see in case it isn't picked up on why files don't list when someone reviews it)

x1arch commented:
@BenjiThatFoxGuy This is a bug of code clearance before the commit, I'll fix it.
for your local testing you can fix it by your self, just replace line 300 in the file `backend/terabox/terabox.go` to next line `if f.root != "/" && f.origRootItem == nil {` and you will get a working ls. Mount I didn't checked yet.

x1arch commented:
@fidodido48 fixed remove and files upload.

x1arch commented:
@fidodido48 do you know, the `rmdir` - `This removes empty directory given by path. Will not remove the path if it has any objects in it, not even empty subdirectories.`

I did not implement this method, because I do not understand why it should be there.... Just use `purge` it can delete files or directories. It's not a rocket science, to make this method, but I do not understand why it should be there. 

I have a lot of questions about rClone commands. And want to make a suggestion in 2.0 version implement  commands from *nix, like `ls` (without recursive scan and it should show all the files and catalogs on requested level), `rm`, `cp` and `mv` which will have a predictable behavior like in Linux, BSD or OS X. This tool mainly for System Admins, whom knows this commands.

mimi0000oo commented:
Hey, I tested out the project right now, using the docs you added to the `backend/terabox/docs/content/terabox.md`.
I didn't manage to find the TSID cookie in the TeraBox website, which seems a bit wired. I've tried using the browserid and  csrfToken instead but this doesn't seem to be anything related to that TSID.

I guess this way it doesn't connect to TeraBox in any way but this is the error:
`[mimi07@archlinux bin]$ ./rclone lsd remote:
panic: runtime error: slice bounds out of range [:1] with length 0

goroutine 1 [running]:
github.com/rclone/rclone/backend/terabox.NewFs({0x400ee80, 0x57e4ca0}, {0x7ffddf75ea87, 0x6}, {0x7ffddf75ea87, 0x0}, {0x3ff8b28, 0xc00057e600})
	/home/mimi07/Desktop/rclone/backend/terabox/terabox.go:174 +0x585
github.com/rclone/rclone/fs.NewFs({0x400ee80, 0x57e4ca0}, {0x7ffddf75ea87, 0x7})
	/home/mimi07/Desktop/rclone/fs/newfs.go:68 +0x33c
github.com/rclone/rclone/fs/cache.GetFn.func1({0x7ffddf75ea87?, 0x100c0000659f8?})
	/home/mimi07/Desktop/rclone/fs/cache/cache.go:102 +0x38
github.com/rclone/rclone/lib/cache.(*Cache).Get(0xc00057e060, {0x7ffddf75ea87, 0x7}, 0xc000065a20)
	/home/mimi07/Desktop/rclone/lib/cache/cache.go:88 +0xe2
github.com/rclone/rclone/fs/cache.GetFn({0x400ee80, 0x57e4ca0}, {0x7ffddf75ea87, 0x7}, 0x3c87c10)
	/home/mimi07/Desktop/rclone/fs/cache/cache.go:101 +0x125
github.com/rclone/rclone/fs/cache.Get({0x400ee80, 0x57e4ca0}, {0x7ffddf75ea87, 0x7})
	/home/mimi07/Desktop/rclone/fs/cache/cache.go:182 +0x7b
github.com/rclone/rclone/cmd.NewFsFile({0x7ffddf75ea87, 0x7})
	/home/mimi07/Desktop/rclone/cmd/cmd.go:92 +0x111
github.com/rclone/rclone/cmd.newFsFileAddFilter({0x7ffddf75ea87, 0x7})
	/home/mimi07/Desktop/rclone/cmd/cmd.go:114 +0x57
github.com/rclone/rclone/cmd.NewFsSrc(...)
	/home/mimi07/Desktop/rclone/cmd/cmd.go:136
github.com/rclone/rclone/cmd/lsd.init.func1(0x55f5b40, {0xc0003d2e80, 0x1, 0x1})
	/home/mimi07/Desktop/rclone/cmd/lsd/lsd.go:60 +0x96
github.com/spf13/cobra.(*Command).execute(0x55f5b40, {0xc0003d2e30, 0x1, 0x1})
	/home/mimi07/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1019 +0xa7b
github.com/spf13/cobra.(*Command).ExecuteC(0x55ec6c0)
	/home/mimi07/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1148 +0x40c
github.com/spf13/cobra.(*Command).Execute(...)
	/home/mimi07/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1071
github.com/rclone/rclone/cmd.Main()
	/home/mimi07/Desktop/rclone/cmd/cmd.go:541 +0x31
main.main()
	/home/mimi07/Desktop/rclone/rclone.go:14 +0xf`

Thank you for your work, i can't wait to try this when it will be released! I'll try to provide more testing for you, let me know if you have any questions.

mimi0000oo commented:
@fidodido48 yea, that's basically the same scenario that i have.

can you please share the cookie that worked for you?

BenjiThatFoxGuy commented:
Hello. To configure the backend I just use https://cookie-editor.com/ on terabox and export as "Header String".
If you account has access to their AI thing, just make sure to click "My Cloud" on the sidebar before exporting the cookies, just in case.
It gives you multiple cookies in a oneliner, paste the whole thing while configuring the backend and it just works.

x1arch commented:
@fidodido48 @mimi0000oo Here is instruction:

To configure an Terabox backend you need provide cookie from your logined account. Three variables need to be provided `TSID`, `ndus`, `ndut_fmt` separated by semicolon.

How to obtain the cookie:
* Chrome extension:
    1) Install extension `Cookie viewer`: https://chromewebstore.google.com/detail/dedhcncdjkmjpebfohadfeeaopiponca?utm_source=item-share-cb
    2) Go to your storage: https://www.terabox.com/main
    3) On search panel exrtensions click to Cookie viewer and copy this three values: `TSID`, `ndus`, `ndut_fmt` insert it one by one to some text document and separate by semicolon, then copy result string for configuration, you should got something like `TSID=xxx; ndus=xxx; ndut_fmt=xxx`
    4) Remove extension `Cookie viewer`

* Chrome dev:
    1) Go to your storage https://www.terabox.com/main
    2) Right top corner `three dots` > `More tools` > `Developer tools` (**Note** The page will be automaticaly redirected to `about:blank`, no worries, it's ok.)
    3) Select `Network` tab and check `Preserve log`
    4) Push to `Go Back` button on address panel
    5) Select any request and scroll `Headers` tab to down under `Request headers` and copy cookie value.
    6) Close dev tools

For Chrome dev variant you can take full cookie string (with other variables), and add it to config.

I have no any problems with download crypted or any other files, I upload and download any files. I didn't check rclone crypt, but I was upload and download crypted/binary/text files without problems.

seiya-git commented:
@fidodido48 probably because of user agent, I use Terabox PC user client:
```
terabox;1.37.0.7;PC;PC-Windows;10.0.22631;WindowsTeraBox
```

seiya-git commented on backend/terabox/terabox.go:
```diff
@@ -0,0 +1,735 @@
+
+const (
+	baseURL       = "https://www.terabox.com"
+	chunkSize     = 4 * fs.Mebi // 4MB
```
chunk sizes is not fixed 4MB if file >4GB

x1arch replied:
Maybe, I don't know. I have no paid account. I can implement only what I have.

seiya-git commented on backend/terabox/terabox.go:
```diff
@@ -0,0 +1,735 @@
+const (
+	baseURL       = "https://www.terabox.com"
+	chunkSize     = 4 * fs.Mebi // 4MB
+	fileLimitSize = 4 * fs.Gibi // 4GB
```
fileLimitSize is 128GB limit for premium users

x1arch replied:
I saw the specs for paid (premium) accounts, but I have no that one.

seiya-git commented on backend/terabox/terabox.go:
```diff
@@ -0,0 +1,735 @@
+			// 	Sensitive: true,
+			// },
+			{
+				Help:     "Set cookie (should contains TSID; ndus; ndut_fmt)",
```
only "ndus" cookie is needed for api

x1arch replied:
`ndus` only not works. Looks like Terabox is checking, when cookie contains two keys and more. `ndus` with `lang` works, but Important thing: `ndus` should be ends with `;`, otherwise it not works, that's mean the cookie string `ndus=%s; lang=en` or , `lang=en; ndus=%s;`  - works, `lang=en; ndus=%s` - not works. 

OK, will update.

seiya-git commented on backend/terabox/terabox.go:
```diff
@@ -0,0 +1,735 @@
+				Help:     "Set custom header User Agent",
+				Name:     "user_agent",
+				Advanced: true,
+				Default:  "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/132.0.0.0 Safari/537.36",
```
better use desktop app user-agent

x1arch replied:
Why? Download speed the same. We use browser API, why desktop UA will be better? Only in case with capcha, for @fidodido48, is it works? The capcha with desktop UA is gone? Lots of time is gone, I believe you already tested it.

seiya-git replied:
Desktop app uses same API as browser but don't have captcha in most cases

seiya-git commented on backend/terabox/api.go:
```diff
@@ -0,0 +1,548 @@
+	opt.MultipartParams.Set("target_path", dirPath)
+
+	if size > int64(chunkSize) {
+		opt.MultipartParams.Set("block_list", `["5910a591dd8fc18c32a8f3df4fdc1761", "a5fc157d78e6ad1c7e114b056c92821e"]`)
```
You should provide actual chunk hashes because it's needed for resuming upload with updated upload_id

seiya-git replied:
Or at least verify the MD5 checksums of uploaded parts and track which parts have been successfully uploaded (I’m not sure if an upload_id can expire, but api should return a list of missing parts if the upload_id is provided in the precreate request along with the actual MD5 hashes of the chunks).

Additionally, the final MD5 can be verified in the createFile request.

x1arch replied:
What?! Sorry man, I don't understand you.

> Or at least verify the MD5 checksums of uploaded chunks

I did it before, BUT, the chunk if it less than 4MB all the time have the different hash than our. I don't know why, because 4MB chunks has the same MD5, but the last one (< 4MB) all the time is different (maybe need to try fill free cells by the 0 under 4MB). BUT the files is OK. I added log of this.

https://github.com/rclone/rclone/blob/9a2c795cde80ffbc8f4dc3f18813526dd67fd83c/backend/terabox/api.go#L426C1-L432C1

> 2025/06/11 16:44:23 DEBUG : Response body: {"md5":"a2fa1e5ba0178ae2ee70584229b8fa02","request_id":3757952266171134165}
2025/06/11 16:44:23 DEBUG : chunk md5 ea70a71e9951acb6ab53952c168e33a7 chunk size 945400
2025/06/11 16:44:23 DEBUG : uploaded chunk have another md5 then our: ea70a71e9951acb6ab53952c168e33a7, uploaded: a2fa1e5ba0178ae2ee70584229b8fa02 will updated

**reMark**: I believe I know why you ask about different `uploadId`, today when I uploading the test file, for to get that log, I got a situation, when file create is failed and uploading starts again. The `uploadId` is changed, but this is absolutely new upload. Chunks, related with `uploadId`, that's why have no reason manipulating with chunks if `uploadId` is different, I checked it. If you try to create file from old chunks, you will get the error `2025/06/11 18:46:53 DEBUG : Response body: {"errmsg":"pcs service failed","errno":31355,"newno":"","request_id":301160865921462432,"server_time":1749667613}`.

>  track which chunks have been successfully uploaded

hmmm... Maybe you're right, I never faced with chunk upload issue, I believe it can save a lot of time if only one chunk broken. Ok, will add three retries for chunk if not got the answer or wrong MD5 for 4MB chunk.

> I’m not sure if an upload_id can expire

Why you thinking about it? Did you faced with `uploadId` expiration? I believe every uploaded chunk reset expiration counter. Any way if you loose your `uploadId`, you can't use the chunks from previous upload. 

> but api should return a list of missing chunks

Not sure about that, but let's imagine it is. And? How it helps us? We upload chunk by chunk and sent result quantity and order of chunks for final file summary why and where we can lose some chunks? If chunk wasn't upload, we return an error and stop uploading. I don't get it.

> Additionally, the final MD5 can be verified in the /api/create request.

We provide all MD5 to this API for create final file from chunks, how we can verify

seiya-git replied:
> I did it before, BUT, the chunk if it less than 4MB all the time have the different hash than our. I don't know why, because 4MB chunks has the same MD5, but the last one (< 4MB) all the time is different (maybe need to try fill free cells by the 0 under 4MB). BUT the files is OK. I added log of this.

Maybe not proper multipart form data? Test it with files under 4MB, it should return same md5

> Not sure about that, but let's imagine it is. And? How it helps us? We upload chunk by chunk and sent result quantity and order of chunks for final file summary why and where we can lose some chunks? If chunk wasn't upload, we return an error and stop uploading. I don't get it.

For files ≤4MB, the encoded MD5 in `/api/create` must match the file's MD5, while for multipart uploads it must match the MD5 of the JSON array containing all chunk MD5s.

For decoding MD5 in `/api/create`:
```go
func decodeMd5(md5 string) string {
	if len(md5) != 32 {
		return md5
	}

	restoredHexChar := fmt.Sprintf("%x", md5[9]-'g')
	o := md5[:9] + restoredHexChar + md5[10:]

	// Apply XOR transformation to each character
	n := make([]byte, 0, 32)
	for i := 0; i < len(o); i++ {
		orig, _ := strconv.ParseInt(string(o[i]), 16, 8)
		xor := int(orig) ^ (i & 15)
		n = append(n, fmt.Sprintf("%x", xor)[0])
	}

	e := string(n[8:16])
		string(n[0:8])
		string(n[24:32])
		string(n[16:24])

	return e
}

x1arch replied:
I found the issue, now the MD5 calculates correct and equal to server MD5. I added check MD5 and reupload chunk if hashes are not equal. Added file reacreate if something went wrong or file create hash not equal our.

x1arch commented:
Branch updated:
- fixed free space reporting (the server returns a free value equal to the total, added manual calculation)
- fixed empty path issue
- updated UserAgent
- updated cookie setting, for now it is allow to provide only `ndus` value
- added reupload chunk if server's MD5 not equal to our MD5
- added retry file create if file create request is failed (faced a few times with SSL handshake timeouts)
- added final hash check when the file upload finished.

x1arch commented:
> But names like `abc:xyz` still don't work, e.g.:

Yeah, I already faced yesterday with this issue, when checking mount, fixed it. It because terabox can't find file in root, if file name not starts from `/`, but can recognize file with path, like `text/ts1500` - works, `ts1500` - not works.

But I have interesting side effect, when I mount terabox backend, the rclone try to start automatically download random file. Yearly, when name without `/` not works, the rclone try download a few files and show you files list, but now it starting download and freezing. Oo

One more thing a lots of errors `net/http: TLS handshake timeout`

x1arch commented:
Branch updated:

- fixed names  which have no leading slash (mount case)
- added `Rmdir`, now `rm` for mount or MC should work @fidodido48 

PS Issue with slow Nautilus works not related with rclone, because MC and console works perfect.

x1arch commented:
@fidodido48 Updated. Checkout please. Is you use driver with encryption or general? I added jsToken to operations, but it's weird, earlier it works without jsToken, but now every second operation request jsToken =(

URenko commented:
> Encryption download do not works

Well, that's a little weird because it works for me.

>  You could try to initiate a new http(s) session for each downloaded file to work around this issue (this will work as a 're-mount/relog').

I actually don't think this is a good idea.
The re-mount getting it to work again might just be because the requests were stalled for a while (the fact that the block removal only lasted for 1–2 files seems to suggest this).

x1arch commented:
@URenko The @fidodido48 use terabox with encryption driver, that's why it's not works, because terabox driver gets from encryption driver filename without the path.  In my case, I use folder `/testenc` as storage for encryption file system, and when terabox gets from encryption driver command to download file, the command if contains only `filename`, when file is located `/testenc/filename`, that's why downloading not works.

@fidodido48 slow down. This is not an API limit, and not related with the sessions and etc. This is related only with downloads methods. Terabox has two methods for downloading the file, the first one is just request file info with full file path, that's main variant of downloading. But in our case it not works, because crypto driver do not provide full path for downloading, we get for this method error `file not found` (-9), when this method is failed, we try download file by the second way, by the file id, and this option is works for you, because we do not need the file path. Earlier download file signature works a long time, the sign keys downloaded ones, when web interface loaded and then this keys uses before page refresh. But for now we need request sign for every download, that's why download for you works only a few minutes.

URenko commented:
I have an account that is not working:
```
2025/06/19 12:47:11 DEBUG : rclone: Version "v1.70.0-beta.8786.1c25362fe.backend/terabox" starting with parameters ["./rclone" "lsd" "__t_M:" "-vvvv"]
2025/06/19 12:47:11 DEBUG : Creating backend with remote "__t_M:"
2025/06/19 12:47:11 DEBUG : Using config file from "……/rclone.conf"
2025/06/19 12:47:11 DEBUG : NewFS __t_M; ; &{Cookie:…… DeletePermanently:false UserAgent:……};
2025/06/19 12:47:15 DEBUG : Request: &{Method:GET URL:https://www.terabox.com/api/check/login?app_id=250528&channel=dubox&clienttype=0 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Accept:[application/json, text/plain, */*] Cookie:……] Referer:[https://www.terabox.com] User-Agent:[……] X-Requested-With:[XMLHttpRequest]] Body:<nil> GetBody:<nil> ContentLength:0 TransferEncoding:[] Close:false Host:www.terabox.com Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr: RequestURI: TLS:<nil> Cancel:<nil> Response:<nil> Pattern: ctx:{emptyCtx:{}} pat:<nil> matches:[] otherValues:map[]}
2025/06/19 12:47:15 DEBUG : Response: &{Status:200 OK StatusCode:200 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Connection:[keep-alive] Content-Type:[application/json; charset=UTF-8] Date:[Thu, 19 Jun 2025 04:47:14 GMT] Flow-Level:[3] Http-X-Isis-Logid:[……] Is-Reg-Region:[false] Logid:[……] Region-Domain-Prefix:[dm] Server:[nginx] Url-Domain-Prefix:[dm] Vary:[Accept-Encoding] Yld:[……] Yme:[^]] Body:0xc000300140 ContentLength:-1 TransferEncoding:[] Close:false Uncompressed:true Trailer:map[] Request:0xc0000cea00 TLS:0xc0001ae000}
2025/06/19 12:47:15 DEBUG : Response body: {"errno":-6,"newno":"","request_id":……,"show_msg":""}
2025/06/19 12:47:15 CRITICAL: Failed to create file system for "__t_M:": Failed to login, please try again later 
```
The source of the problem is that the Terabox account has partitions, and for this account, it will jump to dm.terabox.com instead of www.terabox.com in the browser.
So https://www.terabox.com/api/check/login will return -6 while https://dm.terabox.com/api/check/login is correct.
This can be determined by the `Url-Domain-Prefix` header of the response.

For reference, see my previous fix in alist (the same cookie works in alist): https://github.com/AlistGo/alist/pull/7279

seiya-git commented:
> I have an account that is not working:
> 
> ```
> 2025/06/19 12:47:11 DEBUG : rclone: Version "v1.70.0-beta.8786.1c25362fe.backend/terabox" starting with parameters ["./rclone" "lsd" "__t_M:" "-vvvv"]
> 2025/06/19 12:47:11 DEBUG : Creating backend with remote "__t_M:"
> 2025/06/19 12:47:11 DEBUG : Using config file from "……/rclone.conf"
> 2025/06/19 12:47:11 DEBUG : NewFS __t_M; ; &{Cookie:…… DeletePermanently:false UserAgent:……};
> 2025/06/19 12:47:15 DEBUG : Request: &{Method:GET URL:https://www.terabox.com/api/check/login?app_id=250528&channel=dubox&clienttype=0 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Accept:[application/json, text/plain, */*] Cookie:……] Referer:[https://www.terabox.com] User-Agent:[……] X-Requested-With:[XMLHttpRequest]] Body:<nil> GetBody:<nil> ContentLength:0 TransferEncoding:[] Close:false Host:www.terabox.com Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr: RequestURI: TLS:<nil> Cancel:<nil> Response:<nil> Pattern: ctx:{emptyCtx:{}} pat:<nil> matches:[] otherValues:map[]}
> 2025/06/19 12:47:15 DEBUG : Response: &{Status:200 OK StatusCode:200 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Connection:[keep-alive] Content-Type:[application/json; charset=UTF-8] Date:[Thu, 19 Jun 2025 04:47:14 GMT] Flow-Level:[3] Http-X-Isis-Logid:[……] Is-Reg-Region:[false] Logid:[……] Region-Domain-Prefix:[dm] Server:[nginx] Url-Domain-Prefix:[dm] Vary:[Accept-Encoding] Yld:[……] Yme:[^]] Body:0xc000300140 ContentLength:-1 TransferEncoding:[] Close:false Uncompressed:true Trailer:map[] Request:0xc0000cea00 TLS:0xc0001ae000}
> 2025/06/19 12:47:15 DEBUG : Response body: {"errno":-6,"newno":"","request_id":……,"show_msg":""}
> 2025/06/19 12:47:15 CRITICAL: Failed to create file system for "__t_M:": Failed to login, please try again later 
> ```
> 
> The source of the problem is that the Terabox account has partitions, and for this account, it will jump to dm.terabox.com instead of [www.terabox.com](http://www.terabox.com) in the browser. So https://www.terabox.com/api/check/login will return -6 while https://dm.terabox.com/api/check/login is correct. This can be determined by the `Url-Domain-Prefix` header of the response.
> 
> For reference, see my previous fix in alist (the same cookie works in alist): [AlistGo/alist#7279](https://github.com/AlistGo/alist/pull/7279)

https://jp.terabox.com/api/check/login is works or -6 too?

x1arch commented:
@URenko Updated, check please. Is domain prefix changed for all API requests? or only for a login?

PS I updated for all.

URenko commented:
> @URenko Updated, check please. Is domain prefix changed for all API requests? or only for a login?
> 
> PS I updated for all.

Thank you, it worked. Yes, it's for all API requests.



> https://jp.terabox.com/api/check/login is works or -6 too?

jp.terabox.com/api/check/login does not works for those use dm.terabox.com, but works for those use www.terabox.com

x1arch commented:
Updated:

- add multithread upload

x1arch commented:
@fidodido48 I answered you, https://github.com/rclone/rclone/pull/8508#issuecomment-2984713263 I need a full log for understand what happens there, add `-vvvvv --terabox-debug-level 4` for getting full log. Or wait while I have an opportunity to test it.

deatf100 commented:
> @fidodido48 thanks for the log.
> 
> It's weird, io error happens on different chunks. Just maybe need retry chunks upload on the error. For now we only retry if md5 failed. I reproduced this io error, will try to fix it.
> 
> But in your log I didn't find `use of closed network connection` error.
> 
> > Why does it report file size of 4.381 GiB and then suddenly 3.971Gi?
> 
> I don't know, but I facing with the same thing, sometimes. Anyway it is not affecting anything, because uploaded file have correct size and md5.

@x1arch I try to used it, but i just get timeout errors: "write tcp IP:PORT->IP:PORT: i/o timeout"
It doesn't matter the file, it will give me errors of timeout. About the upload multi-chunks, I get a okish speed 12/14 mb/s of upload, but files just fails to upload. previous ver beta2, I think. was slow, and some time i will get the timeout error, but files did indeed upload to the host.

x1arch commented:
@deatf100 @fidodido48 Please retest, I added three retry for chunk upload.

URenko commented:
Yes, directly using www/dm.terabox also works: https://github.com/URenko/rclone/commit/7a22050d602fc78698b13aabef28aa71dc41c4bb

x1arch commented:
@fidodido48 

Thanks for your investigation (https://github.com/rclone/rclone/issues/8714), but I can't reproduce it. I generate 10k  500Kb files by this script:

```
for i in {1..10000}; do head -c 500K /dev/urandom > "$(date +%s%N)"; done
```

and try to upload to crypted Terabox by the command:

```
./rclone copy local:{source path}/test tbs:/batch --transfers 12 --human-readable --progress --progress-terminal-title --config config -vvvvvvv --rc
```

internal concurrent uploads config was 5, that's mean 5 chunks could be uploaded at one moment, but it is not matter, because files under 4MB upload as one thread.

Here is video of profiler and upload:
-  crypted: https://u.pcloud.link/publink/show?code=XZQ2Xs5ZfO6rXFExk4VTOaYpMuUPBQ88jyDk
- direct: https://u.pcloud.link/publink/show?code=XZD2Xs5Z9S2l5KqhaWLtTRJ0YdvWVmmze6sy

Average heap was 50-78MB, not matter direct upload to Terabox or encrypted Terabox.

Any way you should understand we allocate buffer = `chunk size` * `number of concurrent chunk uploads` * `rclone concurrent uploads`. Ex. if you will upload 12 files in 5 concurrent chunks by 4MB chunk it will about 500MB only for the i/o buffers.

x1arch commented:
@seiya-git @URenko 

- add dynamic chunk for VIP accounts via @seiya-git recommendation
- add location fix by @URenko recommendation
- add VIP checking

Guys could you test VIP functional and if all works as expected, I believe we can publish this backend.

seiya-git commented:
@x1arch Premium file size limit is 128GiB

URenko commented on backend/terabox/terabox.go:
```diff
@@ -0,0 +1,753 @@
+	if opt.UserAgent != "" {
+		clientConfig.UserAgent = opt.UserAgent
+	}
+	clientConfig.Timeout = 5 * fs.Duration(time.Second)
```
Is hardcoding the timeout necessary? This prevents users from customizing it with `--timeout`.
In fact, 5 seconds is too short for some low-quality networks.

George-Seven commented:
Was reluctant about terabox free handling of user data for providing such generous storage, but after testing this pull request, it's great combined with the rclone crypt (amazing), my worries gone.

Got the premium now and tested some relevant features, like uploading files that are more than 4GB, and it's working.

One thing, there's a lot of peer disconnect errors. I had to use gluetun proton vpn in docker, and run this custom rclone fork in it's network namespace and it's working after that. Maybe they don't like the normal isp ip of this region wifi provider?

But the terabox android app works well without vpn on same wifi. Not sure if the app is using a different api than the one on their documentation which may create this distinction.

Also, while it does work, speed isn't as fast as other uploading methods like uploading from the app, but it's okay, machine is running all day with a large cache.

But I had to heavily increase the error retries, because for some reason it errors out too many times, with at least nearly half of the uploads encountering errors. It does eventually go through, which is great. Will probably get ironed out in future updates.

All in all, amazing work! 2TB storage with full zero-knowledge encryption at $15.79/year is simply unbelievable!

emailkanji commented:
> One thing, there's a lot of peer disconnect errors. I had to use gluetun proton vpn in docker, and run this custom rclone fork in it's network namespace and it's working after that. Maybe they don't like the normal isp ip of this region wifi provider?

> But I had to heavily increase the error retries, because for some reason it errors out too many times, with at least nearly half of the uploads encountering errors. It does eventually go through, which is great. Will probably get ironed out in future updates.

This sounds like you've got packet loss. Which wouldn't be surprising for WiFi. It does also sound like rclone/terabox aren't working together as well as they should be.

christaikobo commented:
Terabox is operated by Baidu, and they usually limit it to 80KB/s for non
premium account in China, so 100-500 is rather generous already.

On Thu, Nov 27, 2025, 09:05 fidodido48 ***@***.***> wrote:

> *fidodido48* left a comment (rclone/rclone#8508)
> <https://github.com/rclone/rclone/pull/8508#issuecomment-3585765034>
>
> @fidodido48 <https://github.com/fidodido48> Possibly you got a shadow
> ban, for abuse, wait for a few days and try again or try to get Premium
> account. You have 2 ways: 1 - pay; 2 - play games at Terabox app.
>
> It's already been couple of days and nothing's changed, so i do believe
> it's rather their policy to lure users and then force them to pay. They
> wouldn't be the 1st storage provider to do this. Some of them just operate
> this way.
> Been googling this case a bit and i'm not the only one. I've also tested
> from another free acc. and guess what? Yeah, no speed limit...
>
> Why are you testing a *PREMIUM* acc.? What did you expect - to pay and
> still be limited? ;)
> Just to be clear: upload speed is fine, it's only the DOWNLOAD speed being
> throttled between 100-500kb/s.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rclone/rclone/pull/8508#issuecomment-3585765034>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFY26LEZL5GBTYXSKOEE763363ZK7AVCNFSM6AAAAAB3BOL2ZGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKOBVG43DKMBTGQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

George-Seven commented:
> @George-Seven Thanks, mate! Yes, download is more interesting, because upload works fine, I suppose, but will explore your dump, thanks.

@x1arch I found that httptoolkit in frida mode basically fixes all the the ssl errors, but looks like the downloads is not in HTTPS response -

https://github.com/user-attachments/assets/5253213f-693a-4a72-84e7-0fcbdf6fb185

&nbsp;

It is showing many unparseable.invalid http unsupported protocol errors, which according to similar conversation from the dev, the protocol being used is something else, other than https -

https://github.com/httptoolkit/httptoolkit/issues/608#issuecomment-2257957787

> "Unparseable" and "unsupported HTTP method" (as shown in that screenshot) are basically the same thing really. The traffic shown there is definitely being decrypted successfully (this is not an SSL pinning problem) but it's not HTTP (or anything else closely related) so HTTP Toolkit can't intercept it. You'll struggle to use any similar HTTP proxy tool for this, and your tooling generally to modify/resend these kinds of requests will be limited. [Wireshark](http://wireshark.org/) may be useful though, since it can show the raw bytes without breaking the connection, so you can inspect the traffic, but that's quite complicated to set up for this kind of thing and you'll need to debug the raw traffic byte-by-byte for yourself.

No readable http syntax or content in the logs resemble the upload. It's probably among the the non-http traffic, maybe some custom protocol.

&nbsp;

On the other hand, for upload it is showing very nicely -

<img width="1829" height="809" alt="httptoolkit_terabox_upload_traffic" src="https://github.com/user-attachments/assets/4cae2f62-bdb4-43da-ac1b-a1be5e2a9db7" />

&nbsp;

The dev mentioned wireshark, I'll see if it changes anything.

ameer1234567890 commented:
@x1arch Here is tail of the log file, while it is hanging at 100%.

```
["089f1fc85d7652344c3266c66db33f6d"]
--c2ac691dc5457b401c6d7a599fc23804e4fb7a58cde9021c691ea863736e--
2025/11/29 10:40:55 DEBUG : Response: &{Status:200 OK StatusCode:200 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Connection:[keep-alive] Content-Type:[application/json; charset=UTF-8] Date:[Sat, 29 Nov 2025 05:40:55 GMT] Flow-Level:[3] Logid:[8866160749777434818] Server:[nginx] Vary:[Accept-Encoding] Yld:[8866160749777434818] Yme:[ZIGW+Ss3QE0WaCsHTWniq29Xvv4cVAPxrh5IyiqBmO+5]] Body:0x40011cca80 ContentLength:-1 TransferEncoding:[chunked] Close:false Uncompressed:true Trailer:map[] Request:0x40003fec80 TLS:0x4000aba000}
2025/11/29 10:40:55 DEBUG : Response body: {"errmsg":" file already exist","errno":-8,"newno":"","request_id":8866160749777434818,"server_time":1764394855}
2025/11/29 10:40:55 DEBUG : controlMD5 089f1fc85d7652344c3266c66db33f6d not equal server file md5 , attempt 0
2025/11/29 10:40:55 DEBUG : map[block_list:[["089f1fc85d7652344c3266c66db33f6d"]] local_mtime:[1764357549] path:[/Backup.usb6/Backup.usb8/docker/portainer/portainer.db] size:[262144] target_path:[/Backup.usb6/Backup.usb8/docker/portainer/] uploadid:[N1-NDMuMjMxLjI5Ljc5OjE3NjQzOTQ2MTk6ODg2NjA5NzM1MTQ4NzkwODE1MQ==]]
2025/11/29 10:40:55 DEBUG : Request: &{Method:POST URL:https://www.terabox.com/api/create?app_id=250528&channel=dubox&clienttype=0&isdir=0&jsToken=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX&rtype=0 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Accept:[application/json, text/plain, */*] Content-Type:[multipart/form-data; boundary=ffa334aaa809153ddff4946e05be24d076ea2f7f9c439d8fba522f618f8f] Cookie:[csrfToken=XXXXXXXXXXXXXXXXXXXXXX; browserid=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx; lang=en; __bid_n=XXXXXXXXXXXXXXXXXXXXXXXXXXX; ab_sr=XXXXXXXXXXXXXXXXXXXXXXXXXXXX; __stripe_mid=XXXXXXXXXXXXXXXXXXXXXXXXX; __stripe_sid=XXXXXXXXXXXXXXXXXXXXXXXXXX; ndus=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX; _rdt_uuid=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXx; _ga=XXXXXXXXXXXXXXXXXXXXXXXXXx; _gcl_au=XXXXXXXXXXXXX; _ga_HSVH9T016H=XXXXXXXXXXXXXXXX; _fbp=XXXXXXXXXXXXXXXXXXXXX; ndut_fmt=XXXXXXXXXXXXXXXXXXXX; g_state={"i_l":0,"i_ll":1763292492808,"i_b":"XXXXXXXXXXXXXX"}] Referer:[https://www.terabox.com] User-Agent:[terabox;1.37.0.7;PC;PC-Windows;10.0.22631;WindowsTeraBox] X-Requested-With:[XMLHttpRequest]] Body:{Reader:0x40011ccc40} GetBody:<nil> ContentLength:0 TransferEncoding:[] Close:false Host:www.terabox.com Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr: RequestURI: TLS:<nil> Cancel:<nil> Response:<nil> Pattern: ctx:0x40008ac500 pat:<nil> matches:[] otherValues:map[]}
2025/11/29 10:40:55 DEBUG : Request body: --ffa334aaa809153ddff4946e05be24d076ea2f7f9c439d8fba522f618f8f
Content-Disposition: form-data; name="size"

262144
--ffa334aaa809153ddff4946e05be24d076ea2f7f9c439d8fba522f618f8f
Content-Disposition: form-data; name="target_path"

/Backup.usb6/Backup.usb8/docker/portainer/
--ffa334aaa809153ddff4946e05be24d076ea2f7f9c439d8fba522f618f8f
Content-Disposition: form-data; name="block_list"

["089f1fc85d7652344c3266c66db33f6d"]
--ffa334aaa809153ddff4946e05be24d076ea2f7f9c439d8fba522f618f8f
Content-Disposition: form-data; name="path"

/Backup.usb6/Backup.usb8/docker/portainer/portainer.db
--ffa334aaa809153ddff4946e05be24d076ea2f7f9c439d8fba522f618f8f
Content-Disposition: form-data; name="local_mtime"

1764357549
--ffa334aaa809153ddff4946e05be24d076ea2f7f9c439d8fba522f618f8f
Content-Disposition: form-data; name="uploadid"

N1-NDMuMjMxLjI5Ljc5OjE3NjQzOTQ2MTk6ODg2NjA5NzM1MTQ4NzkwODE1MQ==
--ffa334aaa809153ddff4946e05be24d076ea2f7f9c439d8fba522f618f8f--
2025/11/29 10:40:55 DEBUG : Response: &{Status:200 OK StatusCode:200 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Connection:[keep-alive] Content-Type:[application/json; charset=UTF-8] Date:[Sat, 29 Nov 2025 05:40:55 GMT] Flow-Level:[3] Logid:[8866160813286169028] Server:[nginx] Vary:[Accept-Encoding] Yld:[8866160813286169028] Yme:[ZIGW+Ss3QE0WaisESmriq29Xvv4cVAPwqx5OxSOJmu20Tw==]] Body:0x4000180b00 ContentLength:-1 TransferEncoding:[chunked] Close:false Uncompressed:true Trailer:map[] Request:0x4000551180 TLS:0x4000aba000}
2025/11/29 10:40:55 DEBUG : Response body: {"errmsg":" file already exist","errno":-8,"newno":"","request_id":8866160813286169028,"server_time":1764394855}
2025/11/29 10:40:55 DEBUG : controlMD5 089f1fc85d7652344c3266c66db33f6d not equal server file md5 , attempt 0
2025/11/29 10:40:55 DEBUG : map[block_list:[["089f1fc85d7652344c3266c66db33f6d"]] local_mtime:[1764357549] path:[/Backup.usb6/Backup.usb8/docker/portainer/portainer.db] size:[262144] target_path:[/Backup.usb6/Backup.usb8/docker/portainer/] uploadid:[N1-NDMuMjMxLjI5Ljc5OjE3NjQzOTQ2MTk6ODg2NjA5NzM1MTQ4NzkwODE1MQ==]]
2025/11/29 10:40:55 DEBUG : Request: &{Method:POST URL:https://www.terabox.com/api/create?app_id=250528&channel=dubox&clienttype=0&isdir=0&jsToken=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX&rtype=0 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Accept:[application/json, text/plain, */*] Content-Type:[multipart/form-data; boundary=52db7289fda72f68c915c9c92ff5cd6c3be050c5a97a6ac65d09183212a3] Cookie:[csrfToken=XXXXXXXXXXXXXX; browserid=XXXXXXXXXXXXXXXXXXXXXXXXXX; lang=en; __bid_n=XXXXXXXXXXXXXXXXXXX; ab_sr=XXXXXXXXXXXXXXXXXXXXXXXXXX; __stripe_mid=XXXXXXXXXXXXXXX; __stripe_sid=XXXXXXXXXXXXX; ndus=XXXXXXXXXXXXXXXXXX; _rdt_uuid=XXXXXXXXXXX; _ga=XXXXXXXXXXXXx; _gcl_au=XXXXXXXXXXX; _ga_HSVH9T016H=XXXXXXXXXXXXXXXX; _fbp=XXXXXXXXXXXX; ndut_fmt=XXXXXXXXXXXXXXXXXXXXXX; g_state={"i_l":0,"i_ll":XXXXXXXXXXXX,"i_b":"XXXXXXXXXXXXXXXXXX"}] Referer:[https://www.terabox.com] User-Agent:[terabox;1.37.0.7;PC;PC-Windows;10.0.22631;WindowsTeraBox] X-Requested-With:[XMLHttpRequest]] Body:{Reader:0x4000180cc0} GetBody:<nil> ContentLength:0 TransferEncoding:[] Close:false Host:www.terabox.com Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr: RequestURI: TLS:<nil> Cancel:<nil> Response:<nil> Pattern: ctx:0x40008ac500 pat:<nil> matches:[] otherValues:map[]}
2025/11/29 10:40:55 DEBUG : Request body: --52db7289fda72f68c915c9c92ff5cd6c3be050c5a97a6ac65d09183212a3
Content-Disposition: form-data; name="block_list"

["089f1fc85d7652344c3266c66db33f6d"]
--52db7289fda72f68c915c9c92ff5cd6c3be050c5a97a6ac65d09183212a3
Content-Disposition: form-data; name="path"

/Backup.usb6/Backup.usb8/docker/portainer/portainer.db
--52db7289fda72f68c915c9c92ff5cd6c3be050c5a97a6ac65d09183212a3
Content-Disposition: form-data; name="local_mtime"

1764357549
--52db7289fda72f68c915c9c92ff5cd6c3be050c5a97a6ac65d09183212a3
Content-Disposition: form-data; name="uploadid"

N1-NDMuMjMxLjI5Ljc5OjE3NjQzOTQ2MTk6ODg2NjA5NzM1MTQ4NzkwODE1MQ==
--52db7289fda72f68c915c9c92ff5cd6c3be050c5a97a6ac65d09183212a3
Content-Disposition: form-data; name="size"

262144
--52db7289fda72f68c915c9c92ff5cd6c3be050c5a97a6ac65d09183212a3
Content-Disposition: form-data; name="target_path"

/Backup.usb6/Backup.usb8/docker/portainer/
--52db7289fda72f68c915c9c92ff5cd6c3be050c5a97a6ac65d09183212a3--
2025/11/29 10:40:55 DEBUG : Response: &{Status:200 OK StatusCode:200 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Connection:[keep-alive] Content-Type:[application/json; charset=UTF-8] Date:[Sat, 29 Nov 2025 05:40:55 GMT] Flow-Level:[3] Logid:[8866160881288048296] Server:[nginx] Vary:[Accept-Encoding] Yld:[8866160881288048296] Yme:[ZIGW+Ss3QE0WaCsHTWriq29Xvv4cVAPzqR5MxiqBmOu5]] Body:0x40011ccee0 ContentLength:-1 TransferEncoding:[chunked] Close:false Uncompressed:true Trailer:map[] Request:0x40003ff040 TLS:0x4000aba000}
2025/11/29 10:40:55 DEBUG : Response body: {"errmsg":" file already exist","errno":-8,"newno":"","request_id":8866160881288048296,"server_time":1764394855}
2025/11/29 10:40:55 DEBUG : controlMD5 089f1fc85d7652344c3266c66db33f6d not equal server file md5 , attempt 0
2025/11/29 10:40:55 DEBUG : map[block_list:[["089f1fc85d7652344c3266c66db33f6d"]] local_mtime:[1764357549] path:[/Backup.usb6/Backup.usb8/docker/portainer/portainer.db] size:[262144] target_path:[/Backup.usb6/Backup.usb8/docker/portainer/] uploadid:[N1-NDMuMjMxLjI5Ljc5OjE3NjQzOTQ2MTk6ODg2NjA5NzM1MTQ4NzkwODE1MQ==]]
2025/11/29 10:40:56 DEBUG : Request: &{Method:POST URL:https://www.terabox.com/api/create?app_id=250528&channel=dubox&clienttype=0&isdir=0&jsToken=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX&rtype=0 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Accept:[application/json, text/plain, */*] Content-Type:[multipart/form-data; boundary=ca289f5e592f63e8f5ee68f797fb25349949d80bd0d1de21c8feda9c411c] Cookie:[csrfToken=XXXXXXXXXXXXX; browserid=XXXXXXXXXXXXXXXXXXXXXXXXXX; lang=en; __bid_n=XXXXXXXXXXXX; ab_sr=XXXXXXXXXXXXXXXx; __stripe_mid=XXXXXXXXXXXXXXXX; __stripe_sid=XXXXXXXXXXXXXXXXXX; ndus=XXXXXXXXXXXXXX; _rdt_uuid=XXXXXXXXXXXXXXXXXx; _ga=XXXXXXXXXXXXXX; _gcl_au=XXXXXXXXXXXX; _ga_HSVH9T016H=XXXXXXXXXXXXXXXXXx; _fbp=XXXXXXXXXXXXXX; ndut_fmt=XXXXXXXXXXXXXXXXX; g_state={"i_l":0,"i_ll":XXXXXXXXXXXX,"i_b":"XXXXXXXXXXXXXXXX"}] Referer:[https://www.terabox.com] User-Agent:[terabox;1.37.0.7;PC;PC-Windows;10.0.22631;WindowsTeraBox] X-Requested-With:[XMLHttpRequest]] Body:{Reader:0x4000181260} GetBody:<nil> ContentLength:0 TransferEncoding:[] Close:false Host:www.terabox.com Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr: RequestURI: TLS:<nil> Cancel:<nil> Response:<nil> Pattern: ctx:0x40008ac500 pat:<nil> matches:[] otherValues:map[]}
2025/11/29 10:40:56 DEBUG : Request body: --ca289f5e592f63e8f5ee68f797fb25349949d80bd0d1de21c8feda9c411c
Content-Disposition: form-data; name="uploadid"

N1-NDMuMjMxLjI5Ljc5OjE3NjQzOTQ2MTk6ODg2NjA5NzM1MTQ4NzkwODE1MQ==
--ca289f5e592f63e8f5ee68f797fb25349949d80bd0d1de21c8feda9c411c
Content-Disposition: form-data; name="size"

262144
--ca289f5e592f63e8f5ee68f797fb25349949d80bd0d1de21c8feda9c411c
Content-Disposition: form-data; name="target_path"

/Backup.usb6/Backup.usb8/docker/portainer/
--ca289f5e592f63e8f5ee68f797fb25349949d80bd0d1de21c8feda9c411c
Content-Disposition: form-data; name="block_list"

["089f1fc85d7652344c3266c66db33f6d"]
--ca289f5e592f63e8f5ee68f797fb25349949d80bd0d1de21c8feda9c411c
Content-Disposition: form-data; name="path"

/Backup.usb6/Backup.usb8/docker/portainer/portainer.db
--ca289f5e592f63e8f5ee68f797fb25349949d80bd0d1de21c8feda9c411c
Content-Disposition: form-data; name="local_mtime"

1764357549
--ca289f5e592f63e8f5ee68f797fb25349949d80bd0d1de21c8feda9c411c--
2025/11/29 10:40:56 DEBUG : Response: &{Status:200 OK StatusCode:200 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Connection:[keep-alive] Content-Type:[application/json; charset=UTF-8] Date:[Sat, 29 Nov 2025 05:40:56 GMT] Flow-Level:[3] Logid:[8866160936834104632] Server:[nginx] Vary:[Accept-Encoding] Yld:[8866160936834104632] Yme:[ZIGW+Ss3QE0RYysGTHb+qnFLufkASwb1tAlGySKDmO2xTV8iMWJDNS1H]] Body:0x40006afb60 ContentLength:-1 TransferEncoding:[chunked] Close:false Uncompressed:true Trailer:map[] Request:0x40003ff400 TLS:0x4000aba000}
2025/11/29 10:40:56 DEBUG : Response body: {"errmsg":" file already exist","errno":-8,"newno":"","request_id":8866160936834104632,"server_time":1764394856}
2025/11/29 10:40:56 DEBUG : controlMD5 089f1fc85d7652344c3266c66db33f6d not equal server file md5 , attempt 0
2025/11/29 10:40:56 DEBUG : map[block_list:[["089f1fc85d7652344c3266c66db33f6d"]] local_mtime:[1764357549] path:[/Backup.usb6/Backup.usb8/docker/portainer/portainer.db] size:[262144] target_path:[/Backup.usb6/Backup.usb8/docker/portainer/] uploadid:[N1-NDMuMjMxLjI5Ljc5OjE3NjQzOTQ2MTk6ODg2NjA5NzM1MTQ4NzkwODE1MQ==]]
```

URenko commented on docs/content/overview.md:
```diff
@@ -63,6 +63,7 @@ Here is an overview of the major features of each cloud storage system.
 | SMB                          | -                 | R/W     | Yes              | No              | -         | -        |
 | SugarSync                    | -                 | -       | No               | No              | -         | -        |
 | Storj                        | -                 | R       | No               | No              | -         | -        |
+| Terabox                      | MD5               | DR/W    | No               | No              | -         | -        |
```
The ModTIme is readonly for files, and directories don't have meaningful ModTime. So it should be `R`

x1arch replied:
I theory, we could set timestamp when upload some file.
https://github.com/x1arch/rclone/blob/5241763f1f2b7e4b1c09606e1c6a88d02bfce7c6/backend/terabox/api.go#L670

But, ok, will do as you want.

ameer1234567890 commented:
I am still getting the below error intermittently.
```
error reading destination directory: invalid character '<' looking for beginning of value
```

x1arch commented:
> I am still getting the below error intermittently.
> 
> ```
> error reading destination directory: invalid character '<' looking for beginning of value
> ```

Need more context. Add debugging information. Like command which you try to execute, full output, add `--dump`, `-vvvvv` and etc.

ameer1234567890 commented:
https://github.com/rclone/rclone/pull/8508#issuecomment-3591023094

x1arch commented:
> [#8508 (comment)](https://github.com/rclone/rclone/pull/8508#issuecomment-3591023094)

1. File already exists. This is not a driver issue. You didn't provide a command, I don't know is the file should be overwritten or not.
2. you wrote about 
> I am still getting the below error intermittently.
>
> > error reading destination directory: invalid character '<' looking for beginning of value

I don't see this error there.

Please provide normal debug information.

x1arch commented:
> Can you show the result of the test_all tests please. If it looks good are we ready to merge?

Here: https://github.com/rclone/rclone/pull/8508#issuecomment-3623540475 in Tests section

x1arch commented:
> > Here: [#8508 (comment)](https://github.com/rclone/rclone/pull/8508#issuecomment-3623540475) in Tests section
> 
> Your latest commit is dated Dec. 8, whereas the referenced test execution is dated Dec. 7?

Yea, this was a fix for the very specific (exotic) case, which not covered by the tests. This case only for people from specific locations, most users will never face with this.

I have a request for documentation fix (https://github.com/rclone/rclone/pull/8508#discussion_r2616674091), after that do I need publish a new test result?

Commit:
terabox: Implementation

gablilli commented:
is this pr completely dead? any chance to get it into upstream?

dylanmtaylor commented:
https://github.com/rclone/rclone/actions/runs/23118012108/job/69334806950?pr=8508 shows a TIFF parsing vulnerability, which should be fixed before merging. 

The fix is to update the golang.org/x/image dependency in go.mod:

go get golang.org/x/image@latest

b0rrasca commented:
> @x1arch wrote
> 
> > Is rclone has some storage for temporary data between re-runs? It could to help fix jsToken error.
> 
> We normally store tokens in the config file. You can do this with the `m configmap.Mapper`
> 
> This looks like it is almost ready for merge - would you agree @x1arch ?

Bump! It would be great to see this go live @x1arch

fidodido48 commented:
> @fidodido48 Never faced with permission issue, invalid file name saw a few times in tests, but test restart fix the issue. If you faces with it often, please run your operation with `-vvvvvvv --dump headers --dump responses` additional keys and share dump of terabox requests and responses.

**$ go/bin/rclone.new --version**
```
rclone v1.74.0-beta.9545.d00cbdf49
- os/version: arch (64 bit)
- os/kernel: 6.19.11-zen1-1-zen (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.26.1-X:nodwarf5
- go/linking: dynamic
- go/tags: none

```

My cmd:
**$ go/bin/rclone.new -vvvvvvv --dump headers --dump responses --terabox-upload-threads 14 --transfers 1 --disable DirMove --low-level-retries 3 --ignore-case --filter-from "$RFF2" --log-file DLDIR/var/log/rclone/rclone_crypt_tbox_invalid_fn.log copy "$DLDIR" alltbox_chunker:**

```
2026/04/27 03:33:08 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2026/04/27 03:33:08 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2026/04/27 03:33:08 DEBUG : HTTP REQUEST (req 0x7f1a6689400)
2026/04/27 03:33:08 DEBUG : POST /api/precreate?app_id=250528&channel=dubox&clienttype=0&jsToken=[...] HTTP/1.1
Host: www.terabox.com
User-Agent: terabox;1.37.0.7;PC;PC-Windows;10.0.22631;WindowsTeraBox
Transfer-Encoding: chunked
Accept: application/json, text/plain, */*
Content-Type: multipart/form-data; boundary=bf0c1fc18fff24f90726ec1c67e47b9d7e07f55307c0aeebf651917b74c9
Cookie: [...]
Referer: https://www.terabox.com
X-Requested-With: XMLHttpRequest
Accept-Encoding: gzip

2026/04/27 03:33:08 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2026/04/27 03:33:08 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2026/04/27 03:33:08 DEBUG : HTTP RESPONSE (req 0x7f1a6689400)
2026/04/27 03:33:08 DEBUG : HTTP/1.1 200 OK
Transfer-Encoding: chunked
Connection: keep-alive
Content-Type: application/json; charset=UTF-8
Date: Mon, 27 Apr 2026 02:33:08 GMT
Flow-Level: 3
Logid: 394722016872353341
Server: nginx
Vary: Accept-Encoding
Yld: 394722016872353341
Yme: ZIGW+Ss3QE0XaCsNTHb+qnFLufkASwb0tAFIxSqBmO+4MhBPPHEbbKtH

6d
{"errmsg":"invalid file name","errno":-7,"newno":"","request_id":394722016872353341,"server_time":1777257188}
0

2026/04/27 03:33:08 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2026/04/27 03:33:08 ERROR : Filenamee-2026_04_26_10_24.mkv: Failed to copy: Invalid file name
2026/04/27 03:33:08 ERROR : Attempt 1/3 failed with 1 errors and: Invalid file name

2026/04/27 03:40:59 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2026/04/27 03:40:59 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2026/04/27 03:40:59 DEBUG : HTTP REQUEST (req 0x7f1aa8ac140)
2026/04/27 03:40:59 DEBUG : POST /api/precreate?app_id=250528&channel=dubox&clienttype=0&jsToken=[...] HTTP/1.1
Host: www.terabox.com
User-Agent: terabox;1.37.0.7;PC;PC-Windows;10.0.22631;WindowsTeraBox
Transfer-Encoding: chunked
Accept: application/json, text/plain, */*
Content-Type: multipart/form-data; boundary=72e0e577a819346771a597cee5e55b88be6f90e3155dbb2f2515ba7044b4
Cookie: [...]
Referer: https://www.terabox.com
X-Requested-With: XMLHttpRequest
Accept-Encoding: gzip

2026/04/27 03:40:59 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2026/04/27 03:40:59 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2026/04/27 03:40:59 DEBUG : HTTP RESPONSE (req 0x7f1aa8ac140)
2026/04/27 03:40:59 DEBUG : HTTP/1.1 200 OK
Transfer-Encoding: chunked
Connection: keep-alive
Content-Type: application/json; charset=UTF-8
Date: Mon, 27 Apr 2026 02:40:59 GMT
Flow-Level: 3
Logid: 394848411418868938
Server: nginx
Vary: Accept-Encoding
Yld: 394848411418868938
Yme: ZIGW+Ss3QE0WaCsNTHb+qnFLufkASwbxtAdGySKDmuUKZQ==

6d
{"errmsg":"invalid file name","errno":-7,"newno":"","request_id":394848411418868938,"server_time":1777257659}
0

2026/04/27 03:41:00 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2026/04/27 03:41:00 ERROR : Filenamee-2026_04_26_10_24.mkv: Failed to copy: Invalid file name
2026/04/27 03:41:00 ERROR : Attempt 2/3 failed with 1 errors and: Invalid file name

2026/04/27 03:45:08 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2026/04/27 03:45:08 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2026/04/27 03:45:08 DEBUG : HTTP REQUEST (req 0x7f1aa6477c0)
2026/04/27 03:45:08 DEBUG : POST /api/precreate?app_id=250528&channel=dubox&clienttype=0&jsToken=[...] HTTP/1.1
Host: www.terabox.com
User-Agent: terabox;1.37.0.7;PC;PC-Windows;10.0.22631;WindowsTeraBox
Transfer-Encoding: chunked
Accept: application/json, text/plain, */*
Content-Type: multipart/form-data; boundary=c52d91cf280eafa87550c579b6702b0c4b027f45ac4a6a15ac103c1a87ff
Cookie: [...]
Referer: https://www.terabox.com
X-Requested-With: XMLHttpRequest
Accept-Encoding: gzip

2026/04/27 03:45:08 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2026/04/27 03:45:09 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2026/04/27 03:45:09 DEBUG : HTTP RESPONSE (req 0x7f1aa6477c0)
2026/04/27 03:45:09 DEBUG : HTTP/1.1 200 OK
Transfer-Encoding: chunked
Connection: keep-alive
Content-Type: application/json; charset=UTF-8
Date: Mon, 27 Apr 2026 02:45:08 GMT
Flow-Level: 3
Logid: 394915324387321631
Server: nginx
Vary: Accept-Encoding
Yld: 394915324387321631
Yme: ZIGW+Ss3QE0UYysCSnb+qnFLufkASwb1tAlGySKDmO1+bA==

6d
{"errmsg":"invalid file name","errno":-7,"newno":"","request_id":394915324387321631,"server_time":1777257908}
0

2026/04/27 03:45:09 ERROR : Filenamee-2026_04_26_10_24.mkv: Failed to copy: Invalid file name
2026/04/27 03:45:09 ERROR : Attempt 3/3 failed with 1 errors and: Invalid file name
2026/04/27 03:45:09 INFO  : 
Transferred:   	   11.703 GiB / 11.703 GiB, 100%, 7.973 MiB/s, ETA 0s
Errors:                 1 (retrying may help)
Checks:               117 / 117, 100%, Listed 34201
Elapsed time:     12m11.8s

2026/04/27 03:45:09 DEBUG : 51 go routines active
2026/04/27 03:45:09 NOTICE: Failed to copy: Invalid file name
```

As i said before - this is happening constantly. Tried to upload this one file 7x already and it keeps on failing all the time. Probably tomorrow will succeed... 

According to @ncw post (https://github.com/rclone/rclone/issues/8777#issuecomment-3225003759), this is a proper way of checking a max filename lenght on a remote:

- decrypted
_rclone test info --check-length remote_:

**$ go/bin/rclone.new test info --check-length terabox:**
```
// terabox
maxFileLength = 997 // for 1 byte unicode characters
maxFileLength = 498 // for 2 byte unicode characters
maxFileLength = 332 // for 3 byte unicode characters
maxFileLength = -1 // for 4 byte unicode characters
```

- encrypted:
_rclone test info --check-length remoteCrypt_:

And it crashes here...

**$ go/bin/rclone.new test info --check-length alltbox_chunker:**
```
2026/04/27 03:11:36 NOTICE: EME operates on 1 to 128 block-cipher blocks, you passed 513
panic: EME operates on 1 to 128 block-cipher blocks, you passed 513

goroutine 150 [running]:
log.Panicf({0x4231dc6?, 0x7f9cad695108?}, {0x2d9ea282ecc8?, 0x14a90ee?, 0x2d9ea2b8d808?})
	/usr/lib/go/src/log/log.go:460 +0x74
github.com/rfjakob/eme.Transform({0x4398b20, 0x2d9ea22d3600}, {0x2d9ead593be0, 0x10, 0x10}, {0x2d9ea4e9b500, 0x2010, 0x2a80}, 0x1)
	/home/user/go/pkg/mod/github.com/rfjakob/eme@v1.2.0/eme.go:133 +0x1e5
github.com/rclone/rclone/backend/crypt.(*Cipher).encryptSegment(0x2d9ead593ba0, {0x2d9ea4e8ef00?, 0x0?})
	/home/user/DLDIR/new_rclone_test/rclone/backend/crypt/cipher.go:283 +0x125
github.com/rclone/rclone/backend/crypt.(*Cipher).encryptFileName(0x2d9ead593ba0, {0x2d9ea4e8ef00?, 0x2d9ea5b15cb0?})
	/home/user/DLDIR/new_rclone_test/rclone/backend/crypt/cipher.go:513 +0x27a
github.com/rclone/rclone/backend/crypt.(*Cipher).EncryptFileName(...)
	/home/user/DLDIR/new_rclone_test/rclone/backend/crypt/cipher.go:532
github.com/rclone/rclone/backend/crypt.(*Fs).NewObject(0x2d9eac540b60, {0x43a0880, 0x63f89a0}, {0x2d9ea4e8ef00?, 0x0?})
	/home/user/DLDIR/new_rclone_test/rclone/backend/crypt/crypt.go:487 +0x65
github.com/rclone/rclone/backend/union.(*Fs).NewObject.func1(0x7)
	/home/user/DLDIR/new_rclone_test/rclone/backend/union/union.go:784 +0x87
github.com/rclone/rclone/backend/union.multithread.func1()
	/home/user/DLDIR/new_rclone_test/rclone/backend/union/union.go:1056 +0x4e
created by github.com/rclone/rclone/backend/union.multithread in goroutine 1
	/home/user/DLDIR/new_rclone_test/rclone/backend/union/union.go:1054 +0x47
```

@ncw suggested in some other post to change `filename_encoding=base32` to `base64`, but that would render all my current files unreadable...

oguzhane commented:
incredible work @x1arch looks almost ready to go alive! hope it gets merged to upstream soon

frequently-timesunked commented:
This is one of the most anticipated modules, I don't understand how it's still in limbo. Ugh

christaikobo commented:
The contributor has not been active for a long time now. Someone probably needs to take over, otherwise this is going nowhere.

madeupandprobablydoesnotexist commented:
Alright gang, someone activate their magic code machine
